### PR TITLE
pallet-aura: add feature-flagged explicit slot duration type

### PR DESCRIPTION
**Pull request by rphmeier**
This is feature-flagged for backwards compatibility - runtimes which are intending to upgrade to asynchronous backing should make use of this.

The main motivation of this change is that computing the slot duration with the `MinimumPeriod` doesn't play nicely with asynchronous backing and authoring multiple blocks per slot. `MinimumPeriod` for Cumulus chains doesn't really make a whole lot of sense, as the only clock that they can be compared to is the relay-chain slot number. We will recommend that parachains run with `MinimumPeriod = 0`, so this change is necessary to keep Aura working in that situation.

I added this with the experimental feature flag with LTS in mind.

### CODE REVIEW

**Review by BradleyOlson64 — APPROVED** [review]
I'm guessing this should be included in the User Update Guide?

Enable feature `explicit-slot-duration` on `aura` dependencies.

**Comment by rphmeier** [comment]
> I'm guessing this should be included in the User Update Guide?

Yeah, specifically for the parts where asynchronous backing is actually enabled and more than one block needs to be made per slot.

**Comment by rphmeier** [comment]
Can someone in @paritytech/sdk-node comment on whether it's better to just do the breaking change or to add the feature flag as proposed here?

File: frame/aura/src/lib.rs [hunk 1]
```diff
@@ -60,6 +60,20 @@ pub use pallet::*;
 
 const LOG_TARGET: &str = "runtime::aura";
 
+/// A slot duration provider which infers the slot duration from the
+/// [`pallet_timestamp::Config::MinimumPeriod`] by multiplying it by two, to ensure
+/// that authors have the majority of their slot to author within.
+///
+/// This was the default behavior of the Aura pallet and may be used for
+/// backwards compatibility.
+pub struct MinimumPeriodTimesTwo<T>(sp_std::marker::PhantomData<T>);
```
Comment (liamaharon) [inline frame/aura/src/lib.rs:69]: This doesn't seem to be used anywhere, maybe you meant to use it in `fn slot_duration`?

**Comment by davxy** [comment]
> Can someone in @paritytech/sdk-node comment on whether it's better to just do the breaking change or to add the feature flag as proposed here?

IMO the feature flag is a good way to gently propose a change. However should also be considered that  this breaking change is not "silent". Maybe an explicit comment about the backward compatible default type would be useful and a matter of a very small change for the user... 

Even better would be to use `associated_types_default`. But is unstable :-/

File: frame/aura/src/lib.rs (hunk 1 above)
Comment (rphmeier) [reply frame/aura/src/lib.rs:69]: No, it's strictly necessary with the explicit-slot-duration feature. I don't see a reason this should be feature flagged as it's strictly additive.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: check-dependent-cumulus
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3299673

File: bin/node-template/runtime/src/lib.rs [hunk 2]
```diff
@@ -207,6 +207,7 @@ impl pallet_aura::Config for Runtime {
 	type DisabledValidators = ();
 	type MaxAuthorities = ConstU32<32>;
 	type AllowMultipleBlocksPerSlot = ConstBool<false>;
+	type SlotDuration = pallet_aura::MinimumPeriodTimesTwo<Runtime>;
```
Comment (rphmeier) [inline bin/node-template/runtime/src/lib.rs:210]: I added this just to make the doc-CI happy (as it builds with all workspace features)

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3303040

File: frame/aura/src/lib.rs [hunk 3]
```diff
@@ -95,6 +109,14 @@ pub mod pallet {
 		/// another pallet which enforces some limitation on the number of blocks authors can create
 		/// using the same slot.
 		type AllowMultipleBlocksPerSlot: Get<bool>;
+
+		/// The slot duration Aura should run with, expressed in milliseconds.
+		/// The effective value of this type should not change while the chain is running.
+		///
+		/// This associated type is only present when compiled with the `explicit-slot-duration` flag
+		/// is present.
```
Comment (BradleyOlson64) [inline frame/aura/src/lib.rs]: "is present" looks like a typo

**Comment by rphmeier** [comment]
bot merge

**Review by sam0x17 — APPROVED** [review]

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3327936

File: bin/node-template/runtime/Cargo.toml [hunk 4]
```diff
@@ -16,7 +16,7 @@ targets = ["x86_64-unknown-linux-gnu"]
 codec = { package = "parity-scale-codec", version = "3.6.1", default-features = false, features = ["derive"] }
 scale-info = { version = "2.5.0", default-features = false, features = ["derive"] }
 
-pallet-aura = { version = "4.0.0-dev", default-features = false, path = "../../../frame/aura" }
+pallet-aura = { version = "4.0.0-dev", default-features = false, features = ["explicit-slot-duration"], path = "../../../frame/aura" }
```
Comment (kianenigma) [inline bin/node-template/runtime/Cargo.toml:19]: fwiw if this is more the case of something that is experimental now but you see being useful for everyone, there is also `feature = experimental` already used in a few `frame` based crates.

Comment (rphmeier) [reply bin/node-template/runtime/Cargo.toml:19]: This is exactly what I was looking for. Thanks

**Comment by rphmeier** [comment]
bot merge

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-fmt
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3329497

**Pull request merged**